src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -4,494 +4,502 @@
-            }
+        private int _taggerReferenceCount;
 
-            private async Task SwitchToMainThreadAndHookupWorkspaceAsync()
-            {
-                try
-                {
-                    await _taggerProvider._threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_disposalCancellationSource.Token);
+        public TagComputer(
+            SyntacticClassificationTaggerProvider taggerProvider,
+            ITextBuffer2 subjectBuffer,
+            TimeSpan diffTimeout)
+        {
+            _taggerProvider = taggerProvider;
+            _subjectBuffer = subjectBuffer;
+            _diffTimeout = diffTimeout;
+            _workQueue = new AsyncBatchingWorkQueue<ITextSnapshot>(
+                DelayTimeSpan.NearImmediate,
+                ProcessChangesAsync,
+                equalityComparer: null,
+                taggerProvider._listener,
+                _disposalCancellationSource.Token);
 
-                    // We both try to connect synchronously, and register for workspace registration events.
-                    // It's possible (particularly in tests), to connect in the startup path, but then get a
-                    // previously scheduled, but not yet delivered event.  Don't bother connecting to the
-                    // same workspace again in that case.
-                    var newWorkspace = _workspaceRegistration.Workspace;
-                    if (newWorkspace == _workspace)
-                        return;
+            _lastLineCache = new LastLineCache(taggerProvider._threadingContext);
 
-                    DisconnectFromWorkspace();
+            _workspaceRegistration = Workspace.GetWorkspaceRegistration(subjectBuffer.AsTextContainer());
+            _workspaceRegistration.WorkspaceChanged += OnWorkspaceRegistrationChanged;
 
-                    if (newWorkspace != null)
-                        ConnectToWorkspace(newWorkspace);
-                }
-                catch (OperationCanceledException)
-                {
-                    // can happen if we were disposed of.
-                }
-                catch (Exception e) when (FatalError.ReportAndCatch(e))
-                {
-                    // We were in a fire and forget task.  So just report the NFW and do not let
-                    // this bleed out to an unhandled exception.
-                }
-            }
+            if (_workspaceRegistration.Workspace != null)
+                ConnectToWorkspace(_workspaceRegistration.Workspace);
+
+            _subjectBuffer.ChangedOnBackground += this.OnSubjectBufferChanged;
+        }
 
-            internal void IncrementReferenceCount()
+        public static TagComputer GetOrCreate(
+            SyntacticClassificationTaggerProvider taggerProvider, ITextBuffer2 subjectBuffer)
+        {
+            taggerProvider._threadingContext.ThrowIfNotOnUIThread();
+
+            if (!subjectBuffer.Properties.TryGetProperty(s_uniqueKey, out TagComputer tagComputer))
```

Comment:
>TryGetProperty [](http://example.com/codeflow?start=42&length=14)

why not use GetOrCreateSingletonProperty?

Reply:
Will do!